### ISSUE TITLE
pulumi login on S3 backend with a chosen AWS profile

### ISSUE THREAD

**Issue by ghost** [pulumi/pulumi#9555]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details
Let's say I have 2 AWS accounts configured on my machine and `~/.aws/credentials` look something like this

```
[default]
aws_access_key_id = aa
aws_secret_access_key = bb
region = eu-central-1

[dev]
region = eu-central-1
role_arn = arn:aws:iam::some-role
source_profile = default
```
I create the S3 bucket I want pulumi to use as a backend 
```
aws s3 mb s3://dev-mycompany-pulumi --profile dev
```

but after I run 
```
pulumi login s3://dev-mycompany-pulumi
```

 it will try to find this bucket in `default` account when I try to `pulumi new aws-csharp` and of course break.
```
 AWS_PROFILE=dev pulumi login s3://dev-mycompany-pulumi
```
also doesn't work.

I'd like to specify the aws profile to use when logging in. So something like the following would be nice

`pulumi login s3://dev-mycompany-pulumi --aws-profile dev`

This is a problem for me, cause in our setup we only allow logging into 1 AWS account and from there we assume role to other accounts, but on this account we don't want to keep infrastructure, it's mainly for management.

Maybe it's already possible, but I just couldn't figure it out though?

**Comment by Mejato** [comment]
@ghost  see #4665

**Comment by Mejato** [comment]
![image](https://user-images.githubusercontent.com/41669982/170905134-8b468706-ca61-499d-a956-a43a9cf2383c.png)

**Comment by rkpawaz** [comment]
does not work  if AWS_REGION is set?

**Comment by AaronFriel** [comment]
@rkpawaz does the `s3://...` login URL not work unless the AWS_REGION environment variable is set?

**Comment by rkpawaz** [comment]
the s3://bucket/key?profile=xyz does not work when AWS_REGION is set to a different region. Some aws services set aws region and then this profile (which points to another region) does not work.

**Comment by AaronFriel** [comment]
> Some aws services set aws region and then this profile (which points to another region) does not work.

I see, is it the login command itself that fails, or a Pulumi program following that login when using AWS resources?

**Comment by rkpawaz** [comment]
I used the environment backend url variable. AWS_REGION is the region where I need to create resources. running pulumi up fails when this region is different. in essence the profile is not used. However if I set AWS_PROFILE in cli and run a s3 list it will work. So the bug is probably in how pulumi handles this. Please try this out at your end as my setup is no longer there.

**Comment by AaronFriel** [comment]
The backend / login URL can be entirely different from where resources are provisioned, so I'm not sure if what you identified is a bug.

For our Kubernetes Operator and some user's CI/CD workflows, the S3 bucket a user logs into to store state will be a privileged account, and the target AWS profile would be a target environment account such as prod, or a finer grained account as is often the case.

In other words, I don't think we use the `s3://` URL for `pulumi login` at all during resource provisioning, and this is intended behavior, though it might be worth calling out in our docs.

Does that make sense?

**Comment by rkpawaz** [comment]
"The backend / login URL can be entirely different from where resources are provisioned, so I'm not sure if what you identified is a bug." <--This not working for me is the bug.

eg) I use pulumi with its stack in s3//us-west-1 from a codebuild running in region us-west-2

**Comment by AaronFriel** [comment]
Oh that's interesting. Can you try this?

```shell
pulumi login s3://mybucket/myk?awssdk=v2&region=us-west-1&profile=profile_us-west-1
```

I think that setting `awssdk=v2` and adding the region key ought to work, I think we can update our docs to reflect that if that works.

**Comment by rkpawaz** [comment]
Im sorry, I have moved on from that work so dont have a way to recreate easily.
Also having to give region AND profile does not seem right as region is inferred from profile usually.

**Comment by AaronFriel** [comment]
I think this may be a requirement of the library we're using to access s3 and other blob storage.

**Comment by rkpawaz** [comment]
Can you be provide a link to updated documentation before closing this?

**Comment by AaronFriel** [comment]
@rkpawaz thanks for the reminder! The [State and Backends](https://www.pulumi.com/docs/intro/concepts/state/#logging-into-the-aws-s3-backend) docs will be updated after this is reviewed & merged: https://github.com/pulumi/pulumi-hugo/pull/1870

**Comment by rkpawaz** [comment]
Thank you, i also assume cross account will work as long as iam permissions
allow.

Provisioning resources in aws ac#/regionX, but state is kept in s3 bucket
in  ac#2/regionY.

 I have not tried this pending above issue.

**Comment by AaronFriel** [comment]
@rkpawaz I believe so, if it does not please create a new issue.

**Comment by ekini** [comment]
Hi @rkpawaz , the behaviour of the gocloud.dev library used for the state storage actually conforms to the precedence defined in https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-precedence

Setting the region in the URL I think equals to "Command line options", which is the highest priority. Then comes the env var `AWS_REGION` and only after that settings defined in the config.

On a positive side, you probably don't need a separate profile just for a different region and specify it in the URL.

Also I'm not quite sure the `awssdk=v2` parameter is needed, as the v1 behaves exactly the same (after my preliminary testing).
